pkg/action/util/pod_manager.go

```diff
@@ -0,0 +1,124 @@
+package util
+
+import (
+	"fmt"
+	"github.com/golang/glog"
+	podutil "github.com/turbonomic/kubeturbo/pkg/discovery/util"
+	"github.com/turbonomic/kubeturbo/pkg/turbostore"
+	client "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/typed/core/v1"
+	api "k8s.io/client-go/pkg/api/v1"
+)
+
+type IPodManager interface {
+	GetPodFromDisplayNameOrUUID(kclient *client.Clientset, displayname, uuid string) (*api.Pod, error)
+	CachePod(old, new *api.Pod)
+}
+
+type PodCachedManager struct {
+	podCache turbostore.ITurboCache
+}
+
+func NewPodCachedManager(podCache turbostore.ITurboCache) *PodCachedManager {
+	return &PodCachedManager{
+		podCache: podCache,
+	}
+}
+
+// CachePod caches the pod name and uid with the default expiration duration
+func (p *PodCachedManager) CachePod(old, new *api.Pod) {
+	// Same name or uid is not normal and will not be cached.
+	if old.Name == new.Name || old.UID == new.UID {
+		glog.Errorf("The pods have same name or uid: old(%s,%v), new(%s,%v)", old.Name, old.UID, new.Name, new.UID)
+		return
+	}
+	p.podCache.Set(old.Name, new.Name, 0)
+	p.podCache.Set(string(old.UID), string(new.UID), 0)
+
+	glog.V(4).Infof("Cached pod name and uid: (%s,%v) -> (%s,%v)", old.Name, old.UID, new.Name, new.UID)
+
+	// FIXME: Need also update the entries with value as the old pod
+	// E.g., three actions in a row for the same pod, the third one will get an invalid pod name as the pod has been renamed.
+}
+
+func (p *PodCachedManager) GetPodFromDisplayNameOrUUID(kclient *client.Clientset, displayName, uuid string) (*api.Pod, error) {
+	namespace, name, err := podutil.ParsePodDisplayName(displayName)
+	if err != nil {
+		err = fmt.Errorf("Failed to parse Pod display name %s with uid %s: %v", displayName, uuid, err)
+		glog.Errorf(err.Error())
+		return nil, err
+	}
+
+	pod, err := p.getRunningPod(kclient.CoreV1().Pods(namespace), name, uuid)
+	if err != nil {
+		err = fmt.Errorf("Failed to get Pod by %s/%s: %v", namespace, name, err)
+		glog.Errorf(err.Error())
+		return nil, err
+	}
+
+	return pod, err
+}
+
+func (p *PodCachedManager) getRunningPod(podClient v1.PodInterface, name, uid string) (*api.Pod, error) {
+	// Try the cached name first. If it exists, it means some action was applied so it got renamed.
+	// If try the original name first, it may find the pod that is in terminating process from the previous action
+	// as the action doesn't check if the termination process finishes.
+
+	// Pod is not found with the name. Try using the cached name because actions
+	// might have been applied to the pod.
+	if cachedName, ok := p.getLatestValue(name); ok {
+		glog.V(3).Infof("Using cached name %s for pod %s", cachedName, name)
+		pod, err := podutil.GetPodInPhase(podClient, cachedName.(string), api.PodRunning)
+		if err != nil || pod != nil {
+			return pod, err
+		}
+	}
+
+	pod, err := podutil.GetPodInPhase(podClient, name, api.PodRunning)
```

Comment:
We should wrap GetPodInPhase and GetPodInPhaseByUid in one method, GetPod with both uid and name parsing in, and in GetPod we can try one then the other. This way we prevent the caller to write these procedures again and again.

Reply:
Songbin and I discussed this before. This is intended to try using the pod name for cached first before trying using the uid. The reason is that retrieving pod by uid is more expensive for api server as it caches pod names, instead of uid.